### PR TITLE
C++: Redesign IR dataflow using the shared SSA library

### PR DESCRIPTION
**Pull request by MathiasVP**
Fixes https://github.com/github/codeql-c-team/issues/663.

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll]: Hopefully a75f195df3d50a8cc9d3eb9cef0dbd34aa2491c6 makes things a bit more clear. I <s>stole</s> borrowed your ASCII drawing for the QLDoc.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll]: I've renamed `getAPredecessor` to `getInner` (and similarly for `getASuccesor`) in a75f195df3d50a8cc9d3eb9cef0dbd34aa2491c6. Hopefully, the intuition is more clear now. I'm certainly finding these new names a lot better!

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll]: I've made more liberal use of the `StoreNodeInstr` and `StoreNodeOperand` in a75f195df3d50a8cc9d3eb9cef0dbd34aa2491c6.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll]: d624259eabcf3311017687e208fedb2fbcfb3f43 adds a QLDoc that explains the role of this predicate, and its various cases.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:64]: d34e731f1db5eca97aef475f035e40517a938d61 adds a small novel about the shape of the graph. Hopefully, when combined with a75f195df3d50a8cc9d3eb9cef0dbd34aa2491c6, this should make the code easier to follow.

**Comment by geoffw0** [comment]
SAMATE Juliet test results: https://jenkins.internal.semmle.com/job/Security/job/SAMATE/job/SAMATE-cpp-detailed/183/
Compared to a recent run as baseline: https://jenkins.internal.semmle.com/job/Security/job/SAMATE/job/SAMATE-cpp-detailed/181/
 - headline is that our overall score goes up from 0.083 to 0.088, a decent improvement!
 - we do considerably better at finding TPs on CWE-134 (68% -> 91%, `cpp/tainted-format-string` and `cpp/tainted-format-string-through-global`)
 - we do marginally better at finding TPs on CWE-78 (4% -> 5%, `cpp/command-line-injection`)
 - I don't see any regressions

🎉

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/Ssa.qll
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/Ssa.qll]: Fixed in 092beb8b73cc23c20f568246fb8e6568d51ed608. It took me some time to filter out the special cases where we do want write operations to count as a use, but this change preserves all of our test results. I'll start another DCA to make sure this commit didn't remove any flow.

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/Ssa.qll]: https://github.com/github/codeql-dca-main/issues/1016 compares 092beb8b73cc23c20f568246fb8e6568d51ed608 with the commit _just_ before it. I expect 0 diffs, and potentially some performance wins 🤞.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/Ssa.qll
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/Ssa.qll]: I've added some comments about this in 56cabb8f4609aaf543f36aeef87a90dfba2d572c. Let me know if the disjuncts are still unclear.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/Ssa.qll
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/Ssa.qll]: Good point. 3e6ac74d734428fe5782ded38050b51db164460c adds `InheritanceConversionInstruction` to the list.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll
Comment (jbj) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll:5]: That doesn't sound attractive. Closing this thread.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll
Comment (jbj) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:64]: Thanks!

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll
Comment (jbj) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll:370]: All this seems very special-cased and syntactic. Can it deal with `a = (b ? new A : nullptr)`? What syntax can it not deal with?

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll
Comment (jbj) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll:467]: What's the underlying principle here? How can a future maintainer know whether it's appropriate to add a case for pointer arithmetic, for example? An `InheritanceConversionInstruction` will sometimes do pointer arithmetic.

File: cpp/ql/test/library-tests/dataflow/taint-tests/arrayassignment.cpp
Comment (jbj) [inline cpp/ql/test/library-tests/dataflow/taint-tests/arrayassignment.cpp:57]: Is there anything we can do to restore this result? Would it have worked if pointers were used instead of references?

**Review by jbj — COMMENTED** [review]
I've read all commits now. There's a limit to how deep I've dived into the code, but overall I'm concerned about the many new places where "data flow inside data flow" is used to bridge gaps that are caused by certain syntactic constructs. I'd like to hear your thoughts on whether these sub-data-flow relations can be avoided by exposing more information from the IR.

I'm overall positive about merging this PR because it's shown good results and acceptable performance in testing.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/Ssa.qll
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/Ssa.qll]: Ouch. That change made us to lose a lot of flow. I think I'll revert that commit and get back to this later 😕

File: cpp/ql/test/library-tests/dataflow/taint-tests/map.cpp
Comment (geoffw0) [inline cpp/ql/test/library-tests/dataflow/taint-tests/map.cpp:336]: I wonder if these cases (and the similar cases in `test_pair`) are supposed to be on `m14` rather than `m2`?

**Review by geoffw0 — APPROVED** [review]
The changes to test results look good overall - some regressions, but many improvements as well, and I think quite a lot more of the latter.  The SAMATE and the DCA results I've looked at look good.  I'm also aware of follow-up plans to make things even better.

:+1:

File: cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/ArithmeticUncontrolled/test.c
Comment (MathiasVP) [reply cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/ArithmeticUncontrolled/test.c:100]: Yes. Those are precisely the cases the IR did a good job in helping us detect. I hope we can do https://github.com/github/codeql-c-team/issues/714 as a follow-up to get those results back.

File: cpp/ql/test/library-tests/syntax-zoo/dataflow-ir-consistency.expected
Comment (MathiasVP) [reply cpp/ql/test/library-tests/syntax-zoo/dataflow-ir-consistency.expected:2702]: Yes. All these consistency issues come from the `StoreNodeFlow::flowThrough` predicate which does `PostUpdateNode -> PostUpdateNode` flow to handle all the different kinds of conversions happening between fields lookups. The intention of `PostUpdateNode`s is that they should only arise as the target of a `storeStep`, and should not be produced by `simpleLocalFlowStep`. However, in order to skip these (from the purpose of dataflow) irrelevant dataflow nodes that represent conversions, we need to step over them.

We could get rid of these at the cost of adding another "data flow in data flow" mechanism, but as @jbj alluded to [https://github.com/github/codeql/pull/6825#pullrequestreview-796179710](here) this is probably not a solution we want.

As far as I know, no one is aware of any performance issues caused by having `PostUpdate -> PostUpdate` simple flow steps. The consistency check was just put in since such steps aren't present in the other language's dataflow implementation.

File: cpp/ql/test/library-tests/dataflow/taint-tests/map.cpp
Comment (MathiasVP) [reply cpp/ql/test/library-tests/dataflow/taint-tests/map.cpp:336]: Indeed. It certainly looks weird that there's a bunch of insertions into `m14`, followed by a bunch of lookups on `m2`.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/Ssa.qll
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/Ssa.qll]: ad5619ff072295b1b48d89043e7aa16ad83a4c37 reverts the change. I think we should investigate this as a follow-up.

File: cpp/ql/test/library-tests/dataflow/taint-tests/arrayassignment.cpp
Comment (MathiasVP) [reply cpp/ql/test/library-tests/dataflow/taint-tests/arrayassignment.cpp:57]: I'll take a look at it. I'm sure we can restore it, but I'll probably create an issue for it to not add even more code that needs to be reviewed.

**Comment by MathiasVP** [comment]
> [...] overall I'm concerned about the many new places where "data flow inside data flow" is used to bridge gaps that are caused by certain syntactic constructs. I'd like to hear your thoughts on whether these sub-data-flow relations can be avoided by exposing more information from the IR.

Thanks for going through it all! I share your concern about "data flow inside data flow". I think it would help a lot to make it possible to have conversions "on the side" as we have for AST. The good thing about the IR is that all of these conversions are explicit in the syntax and we're forced to think about them, but the _bad_ thing is that we're forced to handle them everywhere. For instance, with the `valueFlow` predicate that you commented on earlier: it's really just doing the work that a `getUnconvertedInstruction` predicate would give us.

File: cpp/ql/test/library-tests/syntax-zoo/dataflow-ir-consistency.expected
Comment (geoffw0) [reply cpp/ql/test/library-tests/syntax-zoo/dataflow-ir-consistency.expected:2702]: Perhaps we should remove that particular consistency check (for C++)?

Comment (MathiasVP) [reply cpp/ql/test/library-tests/syntax-zoo/dataflow-ir-consistency.expected:2702]: The consistency queries are shared via `identical-files` between all the languages. I can think of a number of ways to handle this:
- Opt-out of `identical-files` for this file
- Add a pyrameterized predicate to modify the behavior of this check on a per-language basis

I'm much more in favor of the second option. We have an instance of this approach already with the  [`isImmutableOrUnobservable`](https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll#L497) predicate. I'm not exactly sure about the interface for this pyrameterized predicate would look like, though.

File: cpp/ql/test/library-tests/dataflow/taint-tests/arrayassignment.cpp
Comment (MathiasVP) [reply cpp/ql/test/library-tests/dataflow/taint-tests/arrayassignment.cpp:57]: The underlying issue is that, after doing the store step for `mi.i = source();` and we end up with flow on the node `VariableAddress[mi] [post update]`, we transfer flow to the read side effect on `mi.get()` (since that is the next load of `mi` following the store to `mi`). But in this case, we actually wanted flow to the `this` argument (and not `*this`) since the read inside `get` looks like:
```
r9_10(glval<unknown>) = VariableAddress[#this]       :
r9_11(MyInt *)        = Load[#this]                  : &:r9_10, m9_6
r9_12(glval<int>)     = FieldAddress[i]              : r9_11
r0_1(int &)           = CopyValue                    : r9_12
m0_2(int &)           = Store[#return]               : &:r9_9, r0_1
```
and we correctly infer the read step from `VariableAddress[#this]` that reads `i`.

Now, this _should_ have been handled by 5dbaea8b52b2eaca2acd71ea4fcee125447538c6, but for some reason it isn't. I _think_ it's not handled because the code returns a reference, and thus the code is looking for a load that's not there. I'll create an issue that describes this in more detail, and we can revisit it later.

One quick fix to this is to delete the ugly hack in 5dbaea8b52b2eaca2acd71ea4fcee125447538c6 and accept object-pointer conflation on call-boundaries, by transfering flow to the `this` argument (instead of `*this`) following store steps. All this should be a one-line change to the `CallInstruction` case in `flowOutOfAddressStep`.

File: cpp/ql/test/library-tests/syntax-zoo/dataflow-ir-consistency.expected
Comment (jbj) [reply cpp/ql/test/library-tests/syntax-zoo/dataflow-ir-consistency.expected:2702]: Before we jump through such hoops, I think you should talk to the other data flow maintainers about whether flow into post-update nodes can sometimes be a good thing. [Last time I checked](https://github.com/github/codeql/pull/3389#issue-610759281), Java could not send post-update flow through casts because it was missing flow into post-update nodes.

Comment (MathiasVP) [reply cpp/ql/test/library-tests/syntax-zoo/dataflow-ir-consistency.expected:2702]: Indeed. It might be that this consistency check should be relaxed _for all languages_.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll:364]: I agree with it being very special-cased and syntactic, yes :(. I hope we can delete this hack with a more clever instantiation to the shared SSA library in the future (that better accounts for indirections).

Regarding `a = (b ? new A : nullptr)`: it works here because the IR inserts a `StoreInstruction` that stores to a temporary after the call to `new A`:

```
r11_12(void *)         = Call[operator new]              : func:r11_10
m11_13(unknown)        = ^CallSideEffect                 : ~m9_4
m11_14(unknown)        = Chi                             : total:m9_4, partial:m11_13
m11_15(unknown)        = ^InitializeDynamicAllocation    : &:r11_12
r11_16(A *)            = Convert                         : r11_12
r11_17(glval<unknown>) = FunctionAddress[A]              :
v11_18(void)           = Call[A]                         : func:r11_17, this:r11_16
m11_19(unknown)        = ^CallSideEffect                 : ~m11_14
m11_20(unknown)        = Chi                             : total:m11_14, partial:m11_19
m11_21(A)              = ^IndirectMayWriteSideEffect[-1] : &:r11_16
m11_22(unknown)        = Chi                             : total:m11_15, partial:m11_21
r11_23(glval<A *>)     = VariableAddress[#temp11:12]     :
m11_24(A *)            = Store[#temp11:12]               : &:r11_23, r11_16
```

So in this case we will flow to `nodeTo.asOperand()` will be the `StoreValueOperand` `r11_16`.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll]: 1f89b4987b76bffecdd4a0574c9508dbf0b0afc2 adds a QLDoc and renames the predicate to better fit the purpose. This is one of those places where it'd be nice to have a "get me the result of `iTo.getUnary()`, but skip past all the conversions.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/Ssa.qll
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/Ssa.qll]: Indeed, it's redundant. Thanks! I've removed this in 43a47952726efc1f424887cab860c6cd6aeeee8c.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/Ssa.qll
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/Ssa.qll]: You're right. I don't like this case either. I _almost_ managed to remove it yesterday. When we don't have this case, we need more flow out of certain `PostUpdateNode`s. I have a local commit that implements this required flow, but the change causes duplicate paths from `argv` in `DefaultTaintTracking`, and I couldn't quite figure out how to solve it in a way that I can confidentially push into this PR.

For now, I've added a comment about it in 4095c2012e214a9e05575f8efdda5f4735dc919c.

**Review by geoffw0 — APPROVED** [review]
I've suggested a couple of things, but none of them should be considered blocking for this PR.

Do any of @jbj s comments require fixes?

File: cpp/ql/test/library-tests/dataflow/taint-tests/map.cpp
Comment (geoffw0) [reply cpp/ql/test/library-tests/dataflow/taint-tests/map.cpp:336]: I think we should fix this as a follow-up (doing it on this PR might make it harder to understand).  Would you like me to do this, or would you rather co-ordinate it yourself?

Comment (MathiasVP) [reply cpp/ql/test/library-tests/dataflow/taint-tests/map.cpp:336]: Good idea. I can do it myself after this has been merged

**Comment by geoffw0** [comment]
We agreed it's probably worth having a slightly deeper look at the [DCA results](https://github.com/github/codeql-dca-main/issues/793) before we merge this.  Here's my bit, by query (I'll come back and add some more next week):

`cpp/cleartext-storage-buffer` (2 new results)
 - both results are in the BAD cases of SAMATE tests

`cpp/command-line-injection` (54 new results)
 - all results are in SAMATE

`cpp/path-injection` (734 new results, 176 lost)
`cpp/tainted-format-string` (432 new results, 24 lost)
`cpp/tainted-format-string-through-global` (96 new results, lost)
`cpp/unbounded-write` (13 new results, 26 lost)
`cpp/uncontrolled-allocation-size` (1 new result)
 - result in Vim, looks like it involves multiple possible paths e.g. from `diff_read`.
 - this would be nice as a path query!

`cpp/uncontrolled-arithmetic` (24 new results, 216 lost)
`cpp/uncontrolled-process-operation` (111 new results, 24 lost)
`cpp/user-controlled-null-termination-tainted` (53 new results, 60 lost)

**Comment by geoffw0** [comment]
Tests are still failing.  I intend to merge this as soon as they're fixed.

**Comment by MathiasVP** [comment]
> Tests are still failing. I intend to merge this as soon as they're fixed.

They're fixed in the internal PR.

**Comment by geoffw0** [comment]
Ah, sorry, I missed that.  I'll review and merge both...

**Review by geoffw0 — APPROVED** [review]
:+1:

Time to merge.

File: cpp/ql/test/library-tests/dataflow/taint-tests/map.cpp
Comment (geoffw0) [reply cpp/ql/test/library-tests/dataflow/taint-tests/map.cpp:336]: (done in https://github.com/github/codeql/pull/7102)

**Pull request merged**
